--- a/arch/arm/mach-cns3xxx/laguna.c
+++ b/arch/arm/mach-cns3xxx/laguna.c
@@ -20,9 +20,11 @@
 #include <linux/init.h>
 #include <linux/kernel.h>
 #include <linux/compiler.h>
+#include <linux/input.h>
 #include <linux/io.h>
 #include <linux/irq.h>
 #include <linux/gpio.h>
+#include <linux/gpio_keys.h>
 #include <linux/dma-mapping.h>
 #include <linux/serial_core.h>
 #include <linux/serial_8250.h>
@@ -612,7 +614,7 @@ static struct at24_platform_data laguna_
 
 static struct pca953x_platform_data laguna_pca_data = {
  	.gpio_base = 100,
-	.irq_base = -1,
+	.irq_base = 168,
 };
 
 static struct pca953x_platform_data laguna_pca2_data = {
@@ -620,15 +622,25 @@ static struct pca953x_platform_data lagu
 	.irq_base = -1,
 };
 
+static struct i2c_board_info gscgpio_i2cinfo = {
+	I2C_BOARD_INFO("pca9555", 0x23),
+	.platform_data = &laguna_pca_data,
+	.irq = 164, /* GSC irq4 */
+};
+
+static struct i2c_board_info pca2_i2cinfo = {
+	I2C_BOARD_INFO("pca9555", 0x27),
+	.platform_data = &laguna_pca2_data,
+};
+
+static struct i2c_board_info gsc_i2cinfo = {
+	I2C_BOARD_INFO("gsc", 0x20),
+	.irq = -1, /* assigned at runtime */
+};
+
 static struct i2c_board_info __initdata laguna_i2c_devices[] = {
 	{
-		I2C_BOARD_INFO("pca9555", 0x23),
-		.platform_data = &laguna_pca_data,
-	},{
-		I2C_BOARD_INFO("pca9555", 0x27),
-		.platform_data = &laguna_pca2_data,
-	},{
-		I2C_BOARD_INFO("gsp", 0x29),
+		I2C_BOARD_INFO("gsc_hwmon", 0x29),
 	},{
 		I2C_BOARD_INFO ("24c08",0x50),
 		.platform_data = &laguna_eeprom_info,
@@ -641,6 +653,11 @@ static struct i2c_board_info __initdata
  * Watchdog
  */
 
+static struct platform_device laguna_gsc_watchdog = {
+	.name	     = "gsc_wdt",
+	.id	     = 0,
+};
+
 static struct resource laguna_watchdog_resources[] = {
 	[0] = {
 		.start	= CNS3XXX_TC11MP_TWD_BASE + 0x100, // CPU0 watchdog
@@ -657,6 +674,14 @@ static struct platform_device laguna_wat
 };
 
 /*
+ * GSC Input device
+ */
+static struct platform_device laguna_gsc_input = {
+	.name		= "gsc_input",
+	.id		= -1,
+};
+
+/*
  * GPS PPS
  */
 static struct pps_gpio_platform_data laguna_pps_data = {
@@ -678,14 +703,12 @@ static struct platform_device laguna_pps
 
 static struct gpio laguna_gpio_gw2391[] = {
 	{   0, GPIOF_IN           , "*GPS_PPS" },
-	{   1, GPIOF_IN           , "*GSC_IRQ#" },
 	{   2, GPIOF_IN           , "*USB_FAULT#" },
 	{   5, GPIOF_OUT_INIT_LOW , "*USB0_PCI_SEL" },
 	{   6, GPIOF_OUT_INIT_HIGH, "*USB_VBUS_EN" },
 	{   7, GPIOF_OUT_INIT_LOW , "*USB1_PCI_SEL" },
 	{   8, GPIOF_OUT_INIT_HIGH, "*PERST#" },
 	{   9, GPIOF_OUT_INIT_LOW , "*FP_SER_EN#" },
-	{ 100, GPIOF_IN           , "*USER_PB#" },
 	{ 103, GPIOF_OUT_INIT_HIGH, "*V5_EN" },
 	{ 108, GPIOF_IN           , "DIO0" },
 	{ 109, GPIOF_IN           , "DIO1" },
@@ -696,13 +719,11 @@ static struct gpio laguna_gpio_gw2391[]
 
 static struct gpio laguna_gpio_gw2388[] = {
 	{   0, GPIOF_IN           , "*GPS_PPS" },
-	{   1, GPIOF_IN           , "*GSC_IRQ#" },
 	{   3, GPIOF_IN           , "*USB_FAULT#" },
 	{   6, GPIOF_OUT_INIT_HIGH, "*USB_VBUS_EN" },
 	{   7, GPIOF_OUT_INIT_LOW , "*GSM_SEL0" },
 	{   8, GPIOF_OUT_INIT_LOW , "*GSM_SEL1" },
 	{   9, GPIOF_OUT_INIT_LOW , "*FP_SER_EN" },
-	{ 100, GPIOF_OUT_INIT_HIGH, "*USER_PB#" },
 	{ 108, GPIOF_IN           , "DIO0" },
 	{ 109, GPIOF_IN           , "DIO1" },
 	{ 110, GPIOF_IN           , "DIO2" },
@@ -712,14 +733,12 @@ static struct gpio laguna_gpio_gw2388[]
 
 static struct gpio laguna_gpio_gw2387[] = {
 	{   0, GPIOF_IN           , "*GPS_PPS" },
-	{   1, GPIOF_IN           , "*GSC_IRQ#" },
 	{   2, GPIOF_IN           , "*USB_FAULT#" },
 	{   5, GPIOF_OUT_INIT_LOW , "*USB_PCI_SEL" },
 	{   6, GPIOF_OUT_INIT_HIGH, "*USB_VBUS_EN" },
 	{   7, GPIOF_OUT_INIT_LOW , "*GSM_SEL0" },
 	{   8, GPIOF_OUT_INIT_LOW , "*GSM_SEL1" },
 	{   9, GPIOF_OUT_INIT_LOW , "*FP_SER_EN" },
-	{ 100, GPIOF_IN           , "*USER_PB#" },
 	{ 103, GPIOF_OUT_INIT_HIGH, "*V5_EN" },
 	{ 108, GPIOF_IN           , "DIO0" },
 	{ 109, GPIOF_IN           , "DIO1" },
@@ -729,8 +748,22 @@ static struct gpio laguna_gpio_gw2387[]
 	{ 113, GPIOF_IN           , "DIO5" },
 };
 
+static struct gpio laguna_gpio_gw2386[] = {
+	{   0, GPIOF_IN           , "*GPS_PPS" },
+	{   2, GPIOF_IN           , "*USB_FAULT#" },
+	{   6, GPIOF_OUT_INIT_LOW , "*USB_PCI_SEL" },
+	{   7, GPIOF_OUT_INIT_LOW , "*GSM_SEL0" },
+	{   8, GPIOF_OUT_INIT_LOW , "*GSM_SEL1" },
+	{   9, GPIOF_OUT_INIT_LOW , "*FP_SER_EN" },
+	{ 108, GPIOF_IN           , "DIO0" },
+	{ 109, GPIOF_IN           , "DIO1" },
+	{ 110, GPIOF_IN           , "DIO2" },
+	{ 111, GPIOF_IN           , "DIO3" },
+	{ 112, GPIOF_IN           , "DIO4" },
+	{ 113, GPIOF_IN           , "DIO5" },
+};
+
 static struct gpio laguna_gpio_gw2385[] = {
-	{   0, GPIOF_IN           , "*GSC_IRQ#" },
 	{   1, GPIOF_OUT_INIT_HIGH, "*USB_HST_VBUS_EN" },
 	{   2, GPIOF_IN           , "*USB_HST_FAULT#" },
 	{   5, GPIOF_IN           , "*USB_OTG_FAULT#" },
@@ -740,12 +773,10 @@ static struct gpio laguna_gpio_gw2385[]
 	{   9, GPIOF_OUT_INIT_LOW , "*SER_EN" },
 	{  10, GPIOF_IN,            "*USER_PB#" },
 	{  11, GPIOF_OUT_INIT_HIGH, "*PERST#" },
-	{ 100, GPIOF_IN           , "*USER_PB#" },
 	{ 103, GPIOF_OUT_INIT_HIGH, "V5_EN" },
 };
 
 static struct gpio laguna_gpio_gw2384[] = {
-	{   0, GPIOF_IN           , "*GSC_IRQ#" },
 	{   1, GPIOF_OUT_INIT_HIGH, "*USB_HST_VBUS_EN" },
 	{   2, GPIOF_IN           , "*USB_HST_FAULT#" },
 	{   5, GPIOF_IN           , "*USB_OTG_FAULT#" },
@@ -757,25 +788,21 @@ static struct gpio laguna_gpio_gw2384[]
 	{  13, GPIOF_OUT_INIT_HIGH, "*I2CMUX_RST#" },
 	{  14, GPIOF_OUT_INIT_LOW , "J10_DIOLED1" },
 	{  15, GPIOF_OUT_INIT_LOW , "J10_DIOLED2" },
-	{ 100, GPIOF_IN           , "*USER_PB#" },
 	{ 103, GPIOF_OUT_INIT_HIGH, "V5_EN" },
 	{ 108, GPIOF_IN           , "J9_DIOGSC0" },
 };
 
 static struct gpio laguna_gpio_gw2383[] = {
 	{   0, GPIOF_IN           , "*GPS_PPS" },
-	{   1, GPIOF_IN           , "*GSC_IRQ#" },
 	{   2, GPIOF_OUT_INIT_HIGH, "*PCIE_RST#" },
 	{   3, GPIOF_IN           , "GPIO0" },
 	{   8, GPIOF_IN           , "GPIO1" },
 	{ 100, GPIOF_IN           , "DIO0" },
 	{ 101, GPIOF_IN           , "DIO1" },
-	{ 108, GPIOF_IN           , "*USER_PB#" },
 };
 
 static struct gpio laguna_gpio_gw2382[] = {
 	{   0, GPIOF_IN           , "*GPS_PPS" },
-	{   1, GPIOF_IN           , "*GSC_IRQ#" },
 	{   2, GPIOF_OUT_INIT_HIGH, "*PCIE_RST#" },
 	{   3, GPIOF_IN           , "GPIO0" },
 	{   4, GPIOF_IN           , "GPIO1" },
@@ -783,19 +810,45 @@ static struct gpio laguna_gpio_gw2382[]
 	{  10, GPIOF_OUT_INIT_HIGH, "*USB_PCI_SEL#" },
 	{ 100, GPIOF_IN           , "DIO0" },
 	{ 101, GPIOF_IN           , "DIO1" },
-	{ 108, GPIOF_IN           , "*USER_PB#" },
 };
 
 static struct gpio laguna_gpio_gw2380[] = {
 	{   0, GPIOF_IN           , "*GPS_PPS" },
-	{   1, GPIOF_IN           , "*GSC_IRQ#" },
 	{   3, GPIOF_IN           , "GPIO0" },
 	{   8, GPIOF_IN           , "GPIO1" },
 	{ 100, GPIOF_IN           , "DIO0" },
 	{ 101, GPIOF_IN           , "DIO1" },
 	{ 102, GPIOF_IN           , "DIO2" },
 	{ 103, GPIOF_IN           , "DIO3" },
-	{ 108, GPIOF_IN           , "*USER_PB#" },
+};
+
+/*
+ * GPIO based buttons
+ */
+static struct gpio_keys_button laguna_keys[] = {
+	{
+		.type			= EV_KEY,
+		.active_low		= 1,
+		.wakeup			= 0,
+		.debounce_interval	= 100,
+		.desc			= "USER_PB#",
+		.code			= BTN_0,
+		.gpio			= -1, /* assigned at runtime */
+	},
+};
+
+static struct gpio_keys_platform_data laguna_keys_pdata = {
+	.buttons = laguna_keys,
+	.nbuttons = ARRAY_SIZE(laguna_keys),
+	.poll_interval = 100,
+};
+
+static struct platform_device laguna_keys_device = {
+	.name = "gpio-keys-polled",
+	.id = 0,
+	.dev = {
+		.platform_data = &laguna_keys_pdata
+	},
 };
 
 /*
@@ -811,10 +864,14 @@ static void __init laguna_init(void)
 				      cns3xxx_cpu_clock() * (1000000 / 8));
 	clk_register_clkdev(clk, "cpu", NULL);
 
-	platform_device_register(&laguna_watchdog);
-
 	platform_device_register(&laguna_i2c_controller);
 
+	/* GPIO controllers */
+	cns3xxx_gpio_init(0, 32, CNS3XXX_GPIOA_BASE_VIRT, IRQ_CNS3XXX_GPIOA,
+			  NR_IRQS_CNS3XXX);
+	cns3xxx_gpio_init(32, 32, CNS3XXX_GPIOB_BASE_VIRT, IRQ_CNS3XXX_GPIOB,
+			  NR_IRQS_CNS3XXX + 32);
+
 	/* Set I2C 0-3 drive strength to 21 mA */
 	reg = MISC_IO_PAD_DRIVE_STRENGTH_CTRL_B;
 	*reg |= 0x300;
@@ -923,8 +980,6 @@ static int __init laguna_model_setup(voi
 		return 0;
 
 	printk("Running on Gateworks Laguna %s\n", laguna_info.model);
-	cns3xxx_gpio_init( 0, 32, CNS3XXX_GPIOA_BASE_VIRT, IRQ_CNS3XXX_GPIOA,
-		NR_IRQS_CNS3XXX);
 
 	/*
 	 * If pcie external interrupts are supported and desired
@@ -941,16 +996,10 @@ static int __init laguna_model_setup(voi
 		reg |= 0x38000000;
 		__raw_writel(reg, mem);
 
-		cns3xxx_gpio_init(32, 32, CNS3XXX_GPIOB_BASE_VIRT,
-				  IRQ_CNS3XXX_GPIOB, NR_IRQS_CNS3XXX + 32);
-
 		irq_set_irq_type(154, IRQ_TYPE_LEVEL_LOW);
 		irq_set_irq_type(93, IRQ_TYPE_LEVEL_HIGH);
 		irq_set_irq_type(94, IRQ_TYPE_LEVEL_HIGH);
 		irq_set_irq_type(95, IRQ_TYPE_LEVEL_HIGH);
-	} else {
-		cns3xxx_gpio_init(32, 32, CNS3XXX_GPIOB_BASE_VIRT,
-				  IRQ_CNS3XXX_GPIOB, NR_IRQS_CNS3XXX + 32);
 	}
 
 	if (strncmp(laguna_info.model, "GW", 2) == 0) {
@@ -1063,19 +1112,56 @@ static int __init laguna_model_setup(voi
 		if ( (strncmp(laguna_info.model, "GW2388", 6) == 0)
 		  || (strncmp(laguna_info.model, "GW2389", 6) == 0) )
 		{
-			// configure GPIO's
+			// GSC
+			laguna_keys[0].gpio = 100;
+			gsc_i2cinfo.irq = 97;
+			i2c_new_device(i2c_get_adapter(0), &gsc_i2cinfo);
+			// GPIO's
 			laguna_register_gpio(ARRAY_AND_SIZE(laguna_gpio_gw2388));
-			// configure LED's
+			// LED's
 			laguna_gpio_leds_data.num_leds = 2;
+
+			i2c_new_device(i2c_get_adapter(0), &gscgpio_i2cinfo);
+			platform_device_register(&laguna_gpio_leds_device);
+			platform_device_register(&laguna_gsc_input);
+			platform_device_register(&laguna_keys_device);
 		} else if (strncmp(laguna_info.model, "GW2387", 6) == 0) {
-			// configure GPIO's
+			// GSC
+			laguna_keys[0].gpio = 100;
+			gsc_i2cinfo.irq = 97;
+			i2c_new_device(i2c_get_adapter(0), &gsc_i2cinfo);
+			// GPIO's
 			laguna_register_gpio(ARRAY_AND_SIZE(laguna_gpio_gw2387));
-			// configure LED's
+			// LED's
+			laguna_gpio_leds_data.num_leds = 2;
+
+			i2c_new_device(i2c_get_adapter(0), &gscgpio_i2cinfo);
+			platform_device_register(&laguna_gpio_leds_device);
+			platform_device_register(&laguna_gsc_input);
+			platform_device_register(&laguna_keys_device);
+		} else if (strncmp(laguna_info.model, "GW2386", 6) == 0) {
+			// GSC
+			laguna_keys[0].gpio = 100;
+			gsc_i2cinfo.irq = 97;
+			i2c_new_device(i2c_get_adapter(0), &gsc_i2cinfo);
+			// GPIO's
+			laguna_register_gpio(ARRAY_AND_SIZE(laguna_gpio_gw2386));
+			// LED's
 			laguna_gpio_leds_data.num_leds = 2;
+
+			i2c_new_device(i2c_get_adapter(0), &gscgpio_i2cinfo);
+			i2c_new_device(i2c_get_adapter(0), &pca2_i2cinfo);
+			platform_device_register(&laguna_gpio_leds_device);
+			platform_device_register(&laguna_gsc_input);
+			platform_device_register(&laguna_keys_device);
 		} else if (strncmp(laguna_info.model, "GW2385", 6) == 0) {
-			// configure GPIO's
+			// GSC
+			laguna_keys[0].gpio = 100;
+			gsc_i2cinfo.irq = 96;
+			i2c_new_device(i2c_get_adapter(0), &gsc_i2cinfo);
+			// GPIO's
 			laguna_register_gpio(ARRAY_AND_SIZE(laguna_gpio_gw2385));
-			// configure LED's
+			// LED's
 			laguna_gpio_leds[0].gpio = 115;
 			laguna_gpio_leds[1].gpio = 12;
 			laguna_gpio_leds[1].name = "red";
@@ -1087,42 +1173,96 @@ static int __init laguna_model_setup(voi
 			laguna_gpio_leds[3].name = "blue";
 			laguna_gpio_leds[3].active_low = 0,
 			laguna_gpio_leds_data.num_leds = 4;
+
+			i2c_new_device(i2c_get_adapter(0), &gscgpio_i2cinfo);
+			platform_device_register(&laguna_gpio_leds_device);
+			platform_device_register(&laguna_gsc_input);
+			platform_device_register(&laguna_keys_device);
 		} else if ( (strncmp(laguna_info.model, "GW2384", 6) == 0)
 			 || (strncmp(laguna_info.model, "GW2394", 6) == 0) )
		{
-			// configure GPIO's
+			// GSC
+			laguna_keys[0].gpio = 100;
+			gsc_i2cinfo.irq = 96;
+			i2c_new_device(i2c_get_adapter(0), &gsc_i2cinfo);
+			// GPIO's
 			laguna_register_gpio(ARRAY_AND_SIZE(laguna_gpio_gw2384));
-			// configure LED's
+			// LED's
 			laguna_gpio_leds_data.num_leds = 1;
+
+			i2c_new_device(i2c_get_adapter(0), &gscgpio_i2cinfo);
+			platform_device_register(&laguna_gpio_leds_device);
+			platform_device_register(&laguna_gsc_input);
+			platform_device_register(&laguna_keys_device);
 		} else if (strncmp(laguna_info.model, "GW2383", 6) == 0) {
-			// configure GPIO's
+			// GSC
+			laguna_keys[0].gpio = 108;
+			gsc_i2cinfo.irq = 97;
+			i2c_new_device(i2c_get_adapter(0), &gsc_i2cinfo);
+			// GPIO's
 			laguna_register_gpio(ARRAY_AND_SIZE(laguna_gpio_gw2383));
-			// configure LED's
+			// LED's
 			laguna_gpio_leds[0].gpio = 107;
 			laguna_gpio_leds_data.num_leds = 1;
+
+			i2c_new_device(i2c_get_adapter(0), &gscgpio_i2cinfo);
+			platform_device_register(&laguna_gpio_leds_device);
+			platform_device_register(&laguna_gsc_input);
+			platform_device_register(&laguna_keys_device);
 		} else if (strncmp(laguna_info.model, "GW2382", 6) == 0) {
-			// configure GPIO's
+			// GSC
+			laguna_keys[0].gpio = 108;
+			gsc_i2cinfo.irq = 97;
+			i2c_new_device(i2c_get_adapter(0), &gsc_i2cinfo);
+			// GPIO's
 			laguna_register_gpio(ARRAY_AND_SIZE(laguna_gpio_gw2382));
-			// configure LED's
+			// LED's
 			laguna_gpio_leds[0].gpio = 107;
 			laguna_gpio_leds_data.num_leds = 1;
+
+			i2c_new_device(i2c_get_adapter(0), &gscgpio_i2cinfo);
+			platform_device_register(&laguna_gpio_leds_device);
+			platform_device_register(&laguna_gsc_input);
+			platform_device_register(&laguna_keys_device);
 		} else if (strncmp(laguna_info.model, "GW2380", 6) == 0) {
-			// configure GPIO's
+			// GSC
+			laguna_keys[0].gpio = 108;
+			gsc_i2cinfo.irq = 97;
+			i2c_new_device(i2c_get_adapter(0), &gsc_i2cinfo);
+			// GPIO's
 			laguna_register_gpio(ARRAY_AND_SIZE(laguna_gpio_gw2380));
-			// configure LED's
+			// LED's
 			laguna_gpio_leds[0].gpio = 107;
 			laguna_gpio_leds[1].gpio = 106;
 			laguna_gpio_leds_data.num_leds = 2;
+
+			i2c_new_device(i2c_get_adapter(0), &gscgpio_i2cinfo);
+			platform_device_register(&laguna_gpio_leds_device);
+			platform_device_register(&laguna_gsc_input);
+			platform_device_register(&laguna_keys_device);
 		} else if ( (strncmp(laguna_info.model, "GW2391", 6) == 0)
 			 || (strncmp(laguna_info.model, "GW2393", 6) == 0) )
 		{
-			// configure GPIO's
+			// GSC
+			laguna_keys[0].gpio = 100;
+			gsc_i2cinfo.irq = 97;
+			i2c_new_device(i2c_get_adapter(0), &gsc_i2cinfo);
+			// GPIO's
 			laguna_register_gpio(ARRAY_AND_SIZE(laguna_gpio_gw2391));
-			// configure LED's
+			// LED's
 			laguna_gpio_leds_data.num_leds = 2;
+
+			platform_device_register(&laguna_gsc_input);
+			i2c_new_device(i2c_get_adapter(0), &gscgpio_i2cinfo);
+			platform_device_register(&laguna_gpio_leds_device);
+			platform_device_register(&laguna_keys_device);
 		}
-		platform_device_register(&laguna_gpio_leds_device);
 	} else {
 		// Do some defaults here, not sure what yet
 	}
+
+	platform_device_register(&laguna_gsc_watchdog);
+
+	platform_device_register(&laguna_watchdog);
+
 	return 0;
 }
 late_initcall(laguna_model_setup);
