--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -295,6 +295,7 @@ dtb-$(CONFIG_SOC_IMX6Q) += \
 	imx6dl-gw552x.dtb \
 	imx6dl-gw553x.dtb \
 	imx6dl-gw5904.dtb \
+	imx6dl-gw5907.dtb \
 	imx6dl-hummingboard.dtb \
 	imx6dl-nit6xlite.dtb \
 	imx6dl-nitrogen6x.dtb \
@@ -321,6 +322,8 @@ dtb-$(CONFIG_SOC_IMX6Q) += \
 	imx6q-gw52xx.dtb \
 	imx6q-gw53xx.dtb \
 	imx6q-gw5400-a.dtb \
+	imx6q-gw5901.dtb \
+	imx6q-gw5902.dtb \
 	imx6q-gw54xx.dtb \
 	imx6q-gw551x.dtb \
 	imx6q-gw552x.dtb \
