Index: linux-4.4/arch/arm/boot/dts/Makefile
===================================================================
--- linux-4.4.orig/arch/arm/boot/dts/Makefile
+++ linux-4.4/arch/arm/boot/dts/Makefile
@@ -296,6 +296,10 @@ dtb-$(CONFIG_SOC_IMX6Q) += \
 	imx6dl-gw553x.dtb \
 	imx6dl-gw560x.dtb \
 	imx6dl-gw5904.dtb \
+	imx6dl-gw5905.dtb \
+	imx6dl-gw5907.dtb \
+	imx6dl-gw5910.dtb \
+	imx6dl-gw5912.dtb \
 	imx6dl-hummingboard.dtb \
 	imx6dl-nit6xlite.dtb \
 	imx6dl-nitrogen6x.dtb \
@@ -322,12 +326,18 @@ dtb-$(CONFIG_SOC_IMX6Q) += \
 	imx6q-gw52xx.dtb \
 	imx6q-gw53xx.dtb \
 	imx6q-gw5400-a.dtb \
+	imx6q-gw5901.dtb \
+	imx6q-gw5902.dtb \
 	imx6q-gw54xx.dtb \
 	imx6q-gw551x.dtb \
 	imx6q-gw552x.dtb \
 	imx6q-gw553x.dtb \
 	imx6q-gw560x.dtb \
 	imx6q-gw5904.dtb \
+	imx6q-gw5905.dtb \
+	imx6q-gw5907.dtb \
+	imx6q-gw5910.dtb \
+	imx6q-gw5912.dtb \
 	imx6q-hummingboard.dtb \
 	imx6q-nitrogen6x.dtb \
 	imx6q-nitrogen6_max.dtb \
